libsolutil/CDCL.cpp

```diff
@@ -0,0 +1,314 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/CDCL.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/StringUtils.h>
+#include <liblangutil/Exceptions.h>
+
+#include <libsolutil/LinearExpression.h>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/tail.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/algorithm/max.hpp>
+#include <range/v3/algorithm/count_if.hpp>
+#include <range/v3/iterator/operations.hpp>
+
+#include <boost/range/algorithm_ext/erase.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::smtutil;
+
+using rational = boost::rational<bigint>;
+
+
+CDCL::CDCL(
+	vector<string> _variables,
+	vector<Clause> const& _clauses,
+	std::function<std::optional<Clause>(std::map<size_t, bool> const&)> _theorySolver
+):
+	m_theorySolver(_theorySolver),
+	m_variables(move(_variables))
+{
+	for (Clause const& clause: _clauses)
+		m_clauses.push_back(make_unique<Clause>(clause));
+
+	for (unique_ptr<Clause> const& clause: m_clauses)
+		setupWatches(*clause);
+
+	// TODO some sanity checks like no empty clauses, no duplicate literals?
+}
+
+optional<CDCL::Model> CDCL::solve()
+{
+	cout << "====" << endl;
+	for (unique_ptr<Clause> const& c: m_clauses)
+		cout << toString(*c) << endl;
+	cout << "====" << endl;
+	while (true)
+	{
+		optional<Clause> conflictClause = propagate();
+		if (!conflictClause && m_theorySolver)
+		{
+			conflictClause = m_theorySolver(m_assignments);
+			if (!conflictClause)
+				cout << "Theory gave us conflict: " << toString(*conflictClause) << endl;
+		}
+		if (conflictClause)
+		{
+			if (currentDecisionLevel() == 0)
+			{
+				cout << "Insatisfiable" << endl;
+				return nullopt;
+			}
+			auto&& [learntClause, backtrackLevel] = analyze(move(*conflictClause));
+			cancelUntil(backtrackLevel);
+			m_clauses.emplace_back(make_unique<Clause>(move(learntClause)));
+			setupWatches(*m_clauses.back());
+		}
+		else
+		{
+			if (auto variable = nextDecisionVariable())
+			{
+				m_decisionPoints.emplace_back(m_assignmentTrail.size());
+				cout << "Deciding on " << m_variables.at(*variable) << " @" << currentDecisionLevel() << endl;
+				enqueue(Literal{false, *variable}, nullptr);
```

Comment:
TODO: we should do polarity caching -- decide on the polarity of the variable it was last set. This is 2-3 lines to do and should bring considerable benefits. (here, we set the polarity fixed, with `Literal{false, *variable}`)

See: https://link.springer.com/chapter/10.1007/978-3-540-72788-0_28